### ISSUE TITLE
Consider a more performant homepage

### ISSUE THREAD

**Issue by madwort** [opensafely-core/opencodelists#1640]
The homepage of this project https://www.opencodelists.org/ currently takes an average of 2.5-3 seconds to load - see [honeycomb bubbleup identifying characteristics of slow pages](https://ui.honeycomb.io/bennett-institute-for-applied-data-science/environments/production/datasets/opencodelists/result/B9k1CdWbAeb?bubbleup_result=3vMkEQWuYAM&tab=bubbleup) and [honeycomb average duration of homepage](https://ui.honeycomb.io/bennett-institute-for-applied-data-science/environments/production/datasets/opencodelists/result/mttLbCiz2jG?hideMarkers). This is also the page that [freshping](https://bennettoxford.freshping.io/reports?check_id=671343) frequently requests to verify the site is online. I can see that there are advantages to having a detailed selection of data on the homepage (for instance, to show off to new visitors, and so that freshping also acts as functional test), but it might also be good for regular visitors to find a snappy, responsive site to get to the information they're looking for quicker & to reduce the load required to run freshping. 

(nb. I consider 'willnotfix' a valid response to this ticket)

**Comment by inglesp** [comment]
Thanks.  We're not going to get to this anytime soon.

**Comment by Jongmassey** [comment]
profiling the [index view](https://github.com/opensafely-core/opencodelists/blob/main/codelists/views/index.py):
```py
>>> import time
>>> start = time.time()
>>> handles = Handle.objects.filter(is_current=True, codelist__is_private=False)
>>> end = time.time()
>>> print(end-start)
0.001859903335571289
>>> start = time.time()
>>> handles = handles.order_by("name")
>>> end = time.time()
>>> print(end-start)
0.0012235641479492188
>>> start = time.time()
>>> codelists = _all_published_codelists(handles)
>>> end = time.time()
>>> print(end-start)
17.353341102600098
```

it's clear that L32 is the culprit.

**Comment by Jongmassey** [comment]
Where there is no search query or `organisation_slug` the index page pulls in all 7384 `Handles` and then iterates them, finding the codelist for each one

```py
def _all_published_codelists(handles):
    return [
        handle.codelist
        for handle in handles
        if handle.codelist.has_published_versions()
    ]
```

These 7384 trips to the database are _quite_ slow, especially when only 15 codelists are displayed on the page.

**Comment by mikerkelly** [comment]
This way of fetching all the data causes a classic N+1 queries. Instead [index.py](https://github.com/opensafely-core/opencodelists/blob/main/codelists/views/index.py), can probably make the DB access into a constant number of queries with a [prefetch_related()](https://docs.djangoproject.com/en/5.1/ref/models/querysets/#prefetch-related) call on the handles QuerySet. Something like (untested):

```python
prefetch= Prefetch('versions', queryset=CodelistVersion.objects.filter(status="published")
handles = Handle.objects.filter(
    is_current=True, codelist__is_private=False).prefetch_related(
    'codelist', codelist_versions_prefetch))
```

Hopefully that would make the query much faster.

If that didn't work sufficiently well, we could either fetch less data, or none, if the search term is blank. Having some on the front page is kind of nice to enable a user to immediately click into one and seeing what the site is about, though. More stats and things would be nice, too... (off topic).

**Comment by Jongmassey** [comment]
Yes, I was also thinking a prefetch of the related codelists would be a good start. It might be worth also doing the pagination calculation before hitting the db, so we only fetch 15 records at a time

**Comment by mikerkelly** [comment]
> Worth noting that the database in question is SQLite, where the N+1 problem [doesn't occur](https://www.sqlite.org/np1queryprob.html) in quite the same way.

Thanks, I wasn't aware of this.

One wrinkle is that each (version of a) coding system is represented in a separate database, for reasons I haven't yet fully understood.

In the course of #2151 I've restored the prod DB to my dev environment but without any coding systems and the homepage or any of the 115 pages of indexed codelists loads in a fraction of a second. This makes me wonder if it's accessing the different DB files that is slowing things down. Maybe we've recreated the N+1 queries problem in SQLite by having multiple databases? Maybe they are repeatedly reopened, for example. I haven't investigated this at all yet.

**Comment by Jongmassey** [comment]
>for reasons I haven't yet fully understood.

https://github.com/opensafely-core/opencodelists/pull/1397 hopefully this is the start of the path to enlightenment

**Comment by mikerkelly** [comment]
Most of this is probably superseded by #2298 and #2299 that we intend to work on in a future opencodelists initiative. It might be worth doing the very quick fix (having an initial search term to reduce the data fetched) sooner (deckscrubbing right now).
